src/main/java/com/learning/pages/PageBase.java

```diff
@@ -1,16 +1,17 @@
 package com.learning.pages;
 
-import org.openqa.selenium.WebDriver;
+import com.learning.browsers.Browser;
 import org.openqa.selenium.support.PageFactory;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Autowired;
 
-public abstract class PageBase {
+public abstract class PageBase implements InitializingBean {
 
-  protected final WebDriver driver;
+  @Autowired
+  protected Browser browser;
 
-  public PageBase(WebDriver driver) {
-    this.driver = driver;
-    PageFactory.initElements(driver, this);
+  @Override
+  public void afterPropertiesSet() {
+    PageFactory.initElements(browser.getWebDriver(), this);
   }
```

Comment:
In this case we have protected `Browser` field with public method to get `WebDriver`. 
So I suppose I did not achieve encapsulation of `WebDriver` object since it can be called by public method right? 
This is detail but quite important I think.

Reply:
Ideally, `getDriver` method should be private, and used only by `Browser`.

Try doing that, change that method to private, and for each occurrence try finding another solution, that would not use driver (ie. use Browser or some other helper instead)